File: tdrs-backend/tdpservice/search_indexes/models/ssp.py
"""Models representing parsed SSP data file records submitted to TDP."""

import uuid
from django.db import models
from tdpservice.data_files.models import DataFile


class SSP_M1(models.Model):
    """
    Parsed record representing an SSP M1 data submission.

    Mapped to an elastic search index.
    """

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    datafile = models.ForeignKey(
        DataFile,
        blank=True,
        help_text='The parent file from which this record was created.',
        null=True,
        on_delete=models.CASCADE,
        related_name='m1_parent'
    )

    RecordType = models.CharField(max_length=156, null=True, blank=False)
    RPT_MONTH_YEAR = models.IntegerField(null=True, blank=False)
    CASE_NUMBER = models.CharField(max_length=11, null=True, blank=False)
    FIPS_CODE = models.CharField(max_length=2, null=True, blank=False)
    COUNTY_FIPS_CODE = models.CharField(
        max_length=3,
        null=True,
        blank=False
    )
    STRATUM = models.CharField(max_length=2, null=True, blank=False)
    ZIP_CODE = models.CharField(max_length=5, null=True, blank=False)
    DISPOSITION = models.IntegerField(null=True, blank=False)
    NBR_FAMILY_MEMBERS = models.IntegerField(null=True, blank=False)
    FAMILY_TYPE = models.IntegerField(null=True, blank=False)
    TANF_ASST_IN_6MONTHS = models.IntegerField(null=True, blank=False)
    RECEIVES_SUB_HOUSING = models.IntegerField(null=True, blank=False)
    RECEIVES_MED_ASSISTANCE = models.IntegerField(null=True, blank=False)
    RECEIVES_FOOD_STAMPS = models.IntegerField(null=True, blank=False)
    AMT_FOOD_STAMP_ASSISTANCE = models.IntegerField(null=True, blank=False)
    RECEIVES_SUB_CC = models.IntegerField(null=True, blank=False)
    AMT_SUB_CC = models.IntegerField(null=True, blank=False)
    CHILD_SUPPORT_AMT = models.IntegerField(null=True, blank=False)
    FAMILY_CASH_RESOURCES = models.IntegerField(null=True, blank=False)
    CASH_AMOUNT = models.IntegerField(null=True, blank=False)
    NBR_MONTHS = models.IntegerField(null=True, blank=False)
    CC_AMOUNT = models.IntegerField(null=True, blank=False)
    CHILDREN_COVERED = models.IntegerField(null=True, blank=False)
    CC_NBR_MONTHS = models.IntegerField(null=True, blank=False)
    TRANSP_AMOUNT = models.IntegerField(null=True, blank=False)
    TRANSP_NBR_MONTHS = models.IntegerField(null=True, blank=False)
    TRANSITION_SERVICES_AMOUNT = models.IntegerField(null=True, blank=False)
    TRANSITION_NBR_MONTHS = models.IntegerField(null=True, blank=False)
    OTHER_AMOUNT = models.IntegerField(null=True, blank=False)
    OTHER_NBR_MONTHS = models.IntegerField(null=True, blank=False)
    SANC_REDUCTION_AMT = models.IntegerField(null=True, blank=False)
    WORK_REQ_SANCTION = models.IntegerField(null=True, blank=False)
    FAMILY_SANC_ADULT = models.IntegerField(null=True, blank=False)
    SANC_TEEN_PARENT = models.IntegerField(null=True, blank=False)
    NON_COOPERATION_CSE = models.IntegerField(null=True, blank=False)
    FAILURE_TO_COMPLY = models.IntegerField(null=True, blank=False)
    OTHER_SANCTION = models.IntegerField(null=True, blank=False)
    RECOUPMENT_PRIOR_OVRPMT = models.IntegerField(null=True, blank=False)
    OTHER_TOTAL_REDUCTIONS = models.IntegerField(null=True, blank=False)
    FAMILY_CAP = models.IntegerField(null=True, blank=False)
    REDUCTIONS_ON_RECEIPTS = models.IntegerField(null=True, blank=False)
    OTHER_NON_SANCTION = models.IntegerField(null=True, blank=False)
    WAIVER_EVAL_CONTROL_GRPS = models.IntegerField(null=True, blank=False)


class SSP_M2(models.Model):
    """
    Parsed record representing an SSP M2 data submission.

    Mapped to an elastic search index.
    """

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    datafile = models.ForeignKey(
        DataFile,
        blank=True,
        help_text='The parent file from which this record was created.',
        null=True,
        on_delete=models.CASCADE,
        related_name='m2_parent'
    )

    RecordType = models.CharField(max_length=156, null=True, blank=False)
    RPT_MONTH_YEAR = models.IntegerField(null=True, blank=False)
    CASE_NUMBER = models.CharField(max_length=11, null=True, blank=False)
    FIPS_CODE = models.CharField(max_length=2, null=True, blank=False)

    FAMILY_AFFILIATION = models.IntegerField(null=True, blank=False)
    NONCUSTODIAL_PARENT = models.IntegerField(null=True, blank=False)
    DATE_OF_BIRTH = models.CharField(max_length=8, null=True, blank=False)
    SSN = models.CharField(max_length=9, null=True, blank=False)
    RACE_HISPANIC = models.IntegerField(null=True, blank=False)
    RACE_AMER_INDIAN = models.IntegerField(null=True, blank=False)
    RACE_ASIAN = models.IntegerField(null=True, blank=False)
    RACE_BLACK = models.IntegerField(null=True, blank=False)
    RACE_HAWAIIAN = models.IntegerField(null=True, blank=False)
    RACE_WHITE = models.IntegerField(null=True, blank=False)
    GENDER = models.IntegerField(null=True, blank=False)
    FED_OASDI_PROGRAM = models.IntegerField(null=True, blank=False)
    FED_DISABILITY_STATUS = models.IntegerField(null=True, blank=False)
    DISABLED_TITLE_XIVAPDT = models.IntegerField(null=True, blank=False)
    AID_AGED_BLIND = models.IntegerField(null=True, blank=False)
    RECEIVE_SSI = models.IntegerField(null=True, blank=False)
    MARITAL_STATUS = models.IntegerField(null=True, blank=False)
    RELATIONSHIP_HOH = models.IntegerField(null=True, blank=False)
    PARENT_MINOR_CHILD = models.IntegerField(null=True, blank=False)
    NEEDS_PREGNANT_WOMAN = models.IntegerField(null=True, blank=False)
    EDUCATION_LEVEL = models.CharField(null=True, blank=False)
    CITIZENSHIP_STATUS = models.IntegerField(null=True, blank=False)
    COOPERATION_CHILD_SUPPORT = models.IntegerField(null=True, blank=False)
    EMPLOYMENT_STATUS = models.IntegerField(null=True, blank=False)
    WORK_ELIGIBLE_INDICATOR = models.IntegerField(null=True, blank=False)
    WORK_PART_STATUS = models.IntegerField(null=True, blank=False)
    UNSUB_EMPLOYMENT = models.IntegerField(null=True, blank=False)
    SUB_PRIVATE_EMPLOYMENT = models.IntegerField(null=True, blank=False)
    SUB_PUBLIC_EMPLOYMENT = models.IntegerField(null=True, blank=False)
    WORK_EXPERIENCE_HOP = models.IntegerField(null=True, blank=False)
    WORK_EXPERIENCE_EA = models.IntegerField(null=True, blank=False)
    WORK_EXPERIENCE_HOL = models.IntegerField(null=True, blank=False)
    OJT = models.IntegerField(null=True, blank=False)
    JOB_SEARCH_HOP = models.IntegerField(null=True, blank=False)
    JOB_SEARCH_EA = models.IntegerField(null=True, blank=False)
    JOB_SEARCH_HOL = models.IntegerField(null=True, blank=False)
    COMM_SERVICES_HOP = models.IntegerField(null=True, blank=False)
    COMM_SERVICES_EA = models.IntegerField(null=True, blank=False)
    COMM_SERVICES_HOL = models.IntegerField(null=True, blank=False)
    VOCATIONAL_ED_TRAINING_HOP = models.IntegerField(null=True, blank=False)
    VOCATIONAL_ED_TRAINING_EA = models.IntegerField(null=True, blank=False)
    VOCATIONAL_ED_TRAINING_HOL = models.IntegerField(null=True, blank=False)
    JOB_SKILLS_TRAINING_HOP = models.IntegerField(null=True, blank=False)
    JOB_SKILLS_TRAINING_EA = models.IntegerField(null=True, blank=False)
    JOB_SKILLS_TRAINING_HOL = models.IntegerField(null=True, blank=False)
    ED_NO_HIGH_SCHOOL_DIPL_HOP = models.IntegerField(null=True, blank=False)
    ED_NO_HIGH_SCHOOL_DIPL_EA = models.IntegerField(null=True, blank=False)
    ED_NO_HIGH_SCHOOL_DIPL_HOL = models.IntegerField(null=True, blank=False)
    SCHOOL_ATTENDENCE_HOP = models.IntegerField(null=True, blank=False)
    SCHOOL_ATTENDENCE_EA = models.IntegerField(null=True, blank=False)
    SCHOOL_ATTENDENCE_HOL = models.IntegerField(null=True, blank=False)
    PROVIDE_CC_HOP = models.IntegerField(null=True, blank=False)
    PROVIDE_CC_EA = models.IntegerField(null=True, blank=False)
    PROVIDE_CC_HOL = models.IntegerField(null=True, blank=False)
    OTHER_WORK_ACTIVITIES = models.IntegerField(null=True, blank=False)
    DEEMED_HOURS_FOR_OVERALL = models.IntegerField(null=True, blank=False)
    DEEMED_HOURS_FOR_TWO_PARENT = models.IntegerField(null=True, blank=False)
    EARNED_INCOME = models.IntegerField(null=True, blank=False)
    UNEARNED_INCOME_TAX_CREDIT = models.IntegerField(null=True, blank=False)
    UNEARNED_SOCIAL_SECURITY = models.IntegerField(null=True, blank=False)
    UNEARNED_SSI = models.IntegerField(null=True, blank=False)
    UNEARNED_WORKERS_COMP = models.IntegerField(null=True, blank=False)
    OTHER_UNEARNED_INCOME = models.IntegerField(null=True, blank=False)


class SSP_M3(models.Model):
    """
    Parsed record representing an SSP M3 data submission.

    Mapped to an elastic search index.
    """

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    datafile = models.ForeignKey(
        DataFile,
        blank=True,
        help_text='The parent file from which this record was created.',
        null=True,
        on_delete=models.CASCADE,
        related_name='m3_parent'
    )

    RecordType = models.CharField(max_length=156, null=True, blank=False)
    RPT_MONTH_YEAR = models.IntegerField(null=True, blank=False)
    CASE_NUMBER = models.CharField(max_length=11, null=True, blank=False)
    FIPS_CODE = models.CharField(max_length=2, null=True, blank=False)

    FAMILY_AFFILIATION = models.IntegerField(null=True, blank=False)
    DATE_OF_BIRTH = models.CharField(max_length=8, null=True, blank=False)
    SSN = models.CharField(max_length=100, null=True, blank=False)
    RACE_HISPANIC = models.IntegerField(null=True, blank=False)
    RACE_AMER_INDIAN = models.IntegerField(null=True, blank=False)
    RACE_ASIAN = models.IntegerField(null=True, blank=False)
    RACE_BLACK = models.IntegerField(null=True, blank=False)
    RACE_HAWAIIAN = models.IntegerField(null=True, blank=False)
    RACE_WHITE = models.IntegerField(null=True, blank=False)
    GENDER = models.IntegerField(null=True, blank=False)
    RECEIVE_NONSSI_BENEFITS = models.IntegerField(null=True, blank=False)
    RECEIVE_SSI = models.IntegerField(null=True, blank=False)
    RELATIONSHIP_HOH = models.IntegerField(null=True, blank=False)
    PARENT_MINOR_CHILD = models.IntegerField(null=True, blank=False)
    EDUCATION_LEVEL = models.CharField(null=True, blank=False)
    CITIZENSHIP_STATUS = models.IntegerField(null=True, blank=False)
    UNEARNED_SSI = models.IntegerField(null=True, blank=False)
    OTHER_UNEARNED_INCOME = models.IntegerField(null=True, blank=False)

class SSP_M4(models.Model):
    """
    Parsed record representing an SSP M1 data submission.

    Mapped to an elastic search index.
    """

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    datafile = models.ForeignKey(
        DataFile,
        blank=True,
        help_text='The parent file from which this record was created.',
        null=True,
        on_delete=models.CASCADE,
        related_name='m4_parent'
    )

    RecordType = models.CharField(max_length=156, null=True, blank=False)
    RPT_MONTH_YEAR = models.IntegerField(null=True, blank=False)
    CASE_NUMBER = models.CharField(max_length=11, null=True, blank=False)
    COUNTY_FIPS_CODE = models.CharField(
        max_length=3,
        null=True,
        blank=False
    )
    STRATUM = models.CharField(max_length=2, null=True, blank=False)
    ZIP_CODE = models.CharField(max_length=5, null=True, blank=False)
    DISPOSITION = models.IntegerField(null=True, blank=False)
    CLOSURE_REASON = models.CharField(max_length=2, null=True, blank=False)
    REC_SUB_HOUSING = models.IntegerField(null=True, blank=False)
    REC_MED_ASSIST = models.IntegerField(null=True, blank=False)
    REC_FOOD_STAMPS = models.IntegerField(null=True, blank=False)
    REC_SUB_CC = models.IntegerField(null=True, blank=False)

class SSP_M5(models.Model):
    """
    Parsed record representing an SSP M1 data submission.

    Mapped to an elastic search index.
    """

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    datafile = models.ForeignKey(
        DataFile,
        blank=True,
        help_text='The parent file from which this record was created.',
        null=True,
        on_delete=models.CASCADE,
        related_name='m5_parent'
    )

    RecordType = models.CharField(max_length=156, null=True, blank=False)
    RPT_MONTH_YEAR = models.IntegerField(null=True, blank=False)
    CASE_NUMBER = models.CharField(max_length=11, null=True, blank=False)

    FAMILY_AFFILIATION = models.IntegerField(null=True, blank=False)
    DATE_OF_BIRTH = models.CharField(max_length=8, null=True, blank=False)
    SSN = models.CharField(max_length=9, null=True, blank=False)
    RACE_HISPANIC = models.IntegerField(null=True, blank=False)
    RACE_AMER_INDIAN = models.IntegerField(null=True, blank=False)
    RACE_ASIAN = models.IntegerField(null=True, blank=False)
    RACE_BLACK = models.IntegerField(null=True, blank=False)
    RACE_HAWAIIAN = models.IntegerField(null=True, blank=False)
    RACE_WHITE = models.IntegerField(null=True, blank=False)
    GENDER = models.IntegerField(null=True, blank=False)
    REC_OASDI_INSURANCE = models.IntegerField(null=True, blank=False)
    REC_FEDERAL_DISABILITY = models.IntegerField(null=True, blank=False)
    REC_AID_TOTALLY_DISABLED = models.IntegerField(null=True, blank=False)
    REC_AID_AGED_BLIND = models.IntegerField(null=True, blank=False)
    REC_SSI = models.IntegerField(null=True, blank=False)
    MARITAL_STATUS = models.IntegerField(null=True, blank=False)
    RELATIONSHIP_HOH = models.CharField(max_length=2, null=True, blank=False)
    PARENT_MINOR_CHILD = models.IntegerField(null=True, blank=False)
    NEEDS_OF_PREGNANT_WOMAN = models.IntegerField(null=True, blank=False)
    EDUCATION_LEVEL = models.CharField(max_length=2, null=True, blank=False)
    CITIZENSHIP_STATUS = models.IntegerField(null=True, blank=False)
    EMPLOYMENT_STATUS = models.IntegerField(null=True, blank=False)
    AMOUNT_EARNED_INCOME = models.CharField(max_length=4, null=True, blank=False)
    AMOUNT_UNEARNED_INCOME = models.CharField(max_length=4, null=True, blank=False)

class SSP_M6(models.Model):
    """
    Parsed record representing an M6 data submission.

    Mapped to an elastic search index.
    """

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    datafile = models.ForeignKey(
        DataFile,
        blank=True,
        help_text='The parent file from which this record was created.',
        null=True,
        on_delete=models.CASCADE,
        related_name='m6_parent'
    )

    RecordType = models.CharField(max_length=156, null=True, blank=False)
    CALENDAR_QUARTER = models.IntegerField(null=True, blank=True)
    RPT_MONTH_YEAR = models.IntegerField(null=True, blank=False)

    SSPMOE_FAMILIES = models.IntegerField(null=True, blank=True)
    NUM_2_PARENTS = models.IntegerField(null=True, blank=True)
    NUM_1_PARENTS = models.IntegerField(null=True, blank=True)
    NUM_NO_PARENTS = models.IntegerField(null=True, blank=True)
    NUM_RECIPIENTS = models.IntegerField(null=True, blank=True)
    ADULT_RECIPIENTS = models.IntegerField(null=True, blank=True)
    CHILD_RECIPIENTS = models.IntegerField(null=True, blank=True)
    NONCUSTODIALS = models.IntegerField(null=True, blank=True)
    AMT_ASSISTANCE = models.IntegerField(null=True, blank=True)
    CLOSED_CASES = models.IntegerField(null=True, blank=True)

class SSP_M7(models.Model):
    """
    Parsed record representing an SSP M3 data submission.

    Mapped to an elastic search index.
    """

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    datafile = models.ForeignKey(
        DataFile,
        blank=True,
        help_text='The parent file from which this record was created.',
        null=True,
        on_delete=models.CASCADE,
        related_name='m7_parent'
    )

    RecordType = models.CharField(max_length=156, null=True, blank=False)
    CALENDAR_QUARTER = models.IntegerField(null=True, blank=True)
    RPT_MONTH_YEAR = models.IntegerField(null=True, blank=False)
    TDRS_SECTION_IND = models.CharField(
        max_length=1,
        null=True,
        blank=False
    )
    STRATUM = models.CharField(max_length=2, null=True, blank=False)
    FAMILIES_MONTH = models.IntegerField(null=True, blank=False)

File: tdrs-services/parser/internal/storage/writer/ssp.go
package writer

import (
	"go-parser/internal/parser"
)

// SSP M1 Serializer - Case-level data for active SSP cases

func serializeSspM1(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		record.Get("RecordType"),
		record.Get("RPT_MONTH_YEAR"),
		record.Get("CASE_NUMBER"),
		record.Get("COUNTY_FIPS_CODE"),
		record.Get("STRATUM"),
		record.Get("ZIP_CODE"),
		record.Get("DISPOSITION"),
		record.Get("NBR_FAMILY_MEMBERS"),
		record.Get("FAMILY_TYPE"),
		record.Get("TANF_ASST_IN_6MONTHS"),
		record.Get("RECEIVES_SUB_HOUSING"),
		record.Get("RECEIVES_MED_ASSISTANCE"),
		record.Get("RECEIVES_FOOD_STAMPS"),
		record.Get("AMT_FOOD_STAMP_ASSISTANCE"),
		record.Get("RECEIVES_SUB_CC"),
		record.Get("AMT_SUB_CC"),
		record.Get("CHILD_SUPPORT_AMT"),
		record.Get("FAMILY_CASH_RESOURCES"),
		record.Get("CASH_AMOUNT"),
		record.Get("NBR_MONTHS"),
		record.Get("CC_AMOUNT"),
		record.Get("CHILDREN_COVERED"),
		record.Get("CC_NBR_MONTHS"),
		record.Get("TRANSP_AMOUNT"),
		record.Get("TRANSP_NBR_MONTHS"),
		record.Get("TRANSITION_SERVICES_AMOUNT"),
		record.Get("TRANSITION_NBR_MONTHS"),
		record.Get("OTHER_AMOUNT"),
		record.Get("OTHER_NBR_MONTHS"),
		record.Get("SANC_REDUCTION_AMT"),
		record.Get("WORK_REQ_SANCTION"),
		record.Get("FAMILY_SANC_ADULT"),
		record.Get("SANC_TEEN_PARENT"),
		record.Get("NON_COOPERATION_CSE"),
		record.Get("FAILURE_TO_COMPLY"),
		record.Get("OTHER_SANCTION"),
		record.Get("RECOUPMENT_PRIOR_OVRPMT"),
		record.Get("OTHER_TOTAL_REDUCTIONS"),
		record.Get("FAMILY_CAP"),
		record.Get("REDUCTIONS_ON_RECEIPTS"),
		record.Get("OTHER_NON_SANCTION"),
		record.Get("WAIVER_EVAL_CONTROL_GRPS"),
		newUUID(),
		datafileID,
		int32(record.LineNumber),
	})
}

// SSP M2 Serializer - Adult-level data for active SSP cases

func serializeSspM2(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		record.Get("RecordType"),
		record.Get("RPT_MONTH_YEAR"),
		record.Get("CASE_NUMBER"),
		record.Get("FIPS_CODE"),
		record.Get("FAMILY_AFFILIATION"),
		record.Get("NONCUSTODIAL_PARENT"),
		record.Get("DATE_OF_BIRTH"),
		record.Get("SSN"),
		record.Get("RACE_HISPANIC"),
		record.Get("RACE_AMER_INDIAN"),
		record.Get("RACE_ASIAN"),
		record.Get("RACE_BLACK"),
		record.Get("RACE_HAWAIIAN"),
		record.Get("RACE_WHITE"),
		record.Get("SEX"),
		record.Get("FED_OASDI_PROGRAM"),
		record.Get("FED_DISABILITY_STATUS"),
		record.Get("DISABLED_TITLE_XIVAPDT"),
		record.Get("AID_AGED_BLIND"),
		record.Get("RECEIVE_SSI"),
		record.Get("MARITAL_STATUS"),
		record.Get("RELATIONSHIP_HOH"),
		record.Get("PARENT_MINOR_CHILD"),
		record.Get("NEEDS_PREGNANT_WOMAN"),
		record.Get("EDUCATION_LEVEL"),
		record.Get("CITIZENSHIP_STATUS"),
		record.Get("COOPERATION_CHILD_SUPPORT"),
		record.Get("EMPLOYMENT_STATUS"),
		record.Get("WORK_ELIGIBLE_INDICATOR"),
		record.Get("WORK_PART_STATUS"),
		record.Get("UNSUB_EMPLOYMENT"),
		record.Get("SUB_PRIVATE_EMPLOYMENT"),
		record.Get("SUB_PUBLIC_EMPLOYMENT"),
		record.Get("WORK_EXPERIENCE_HOP"),
		record.Get("WORK_EXPERIENCE_EA"),
		record.Get("WORK_EXPERIENCE_HOL"),
		record.Get("OJT"),
		record.Get("JOB_SEARCH_HOP"),
		record.Get("JOB_SEARCH_EA"),
		record.Get("JOB_SEARCH_HOL"),
		record.Get("COMM_SERVICES_HOP"),
		record.Get("COMM_SERVICES_EA"),
		record.Get("COMM_SERVICES_HOL"),
		record.Get("VOCATIONAL_ED_TRAINING_HOP"),
		record.Get("VOCATIONAL_ED_TRAINING_EA"),
		record.Get("VOCATIONAL_ED_TRAINING_HOL"),
		record.Get("JOB_SKILLS_TRAINING_HOP"),
		record.Get("JOB_SKILLS_TRAINING_EA"),
		record.Get("JOB_SKILLS_TRAINING_HOL"),
		record.Get("ED_NO_HIGH_SCHOOL_DIPL_HOP"),
		record.Get("ED_NO_HIGH_SCHOOL_DIPL_EA"),
		record.Get("ED_NO_HIGH_SCHOOL_DIPL_HOL"),
		record.Get("SCHOOL_ATTENDENCE_HOP"),
		record.Get("SCHOOL_ATTENDENCE_EA"),
		record.Get("SCHOOL_ATTENDENCE_HOL"),
		record.Get("PROVIDE_CC_HOP"),
		record.Get("PROVIDE_CC_EA"),
		record.Get("PROVIDE_CC_HOL"),
		record.Get("OTHER_WORK_ACTIVITIES"),
		record.Get("DEEMED_HOURS_FOR_OVERALL"),
		record.Get("DEEMED_HOURS_FOR_TWO_PARENT"),
		record.Get("EARNED_INCOME"),
		record.Get("UNEARNED_INCOME_TAX_CREDIT"),
		record.Get("UNEARNED_SOCIAL_SECURITY"),
		record.Get("UNEARNED_SSI"),
		record.Get("UNEARNED_WORKERS_COMP"),
		record.Get("OTHER_UNEARNED_INCOME"),
		newUUID(),
		datafileID,
		int32(record.LineNumber),
	})
}

// SSP M3 Serializer - Child-level data for active SSP cases

func serializeSspM3(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		record.Get("RecordType"),
		record.Get("RPT_MONTH_YEAR"),
		record.Get("CASE_NUMBER"),
		record.Get("FAMILY_AFFILIATION"),
		record.Get("DATE_OF_BIRTH"),
		record.Get("SSN"),
		record.Get("RACE_HISPANIC"),
		record.Get("RACE_AMER_INDIAN"),
		record.Get("RACE_ASIAN"),
		record.Get("RACE_BLACK"),
		record.Get("RACE_HAWAIIAN"),
		record.Get("RACE_WHITE"),
		record.Get("SEX"),
		record.Get("RECEIVE_NONSSI_BENEFITS"),
		record.Get("RECEIVE_SSI"),
		record.Get("RELATIONSHIP_HOH"),
		record.Get("PARENT_MINOR_CHILD"),
		record.Get("EDUCATION_LEVEL"),
		record.Get("CITIZENSHIP_STATUS"),
		record.Get("UNEARNED_SSI"),
		record.Get("OTHER_UNEARNED_INCOME"),
		newUUID(),
		datafileID,
		int32(record.LineNumber),
	})
}

// SSP M4 Serializer - Case-level data for closed SSP cases

func serializeSspM4(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		record.Get("RecordType"),
		record.Get("RPT_MONTH_YEAR"),
		record.Get("CASE_NUMBER"),
		record.Get("COUNTY_FIPS_CODE"),
		record.Get("STRATUM"),
		record.Get("ZIP_CODE"),
		record.Get("DISPOSITION"),
		record.Get("CLOSURE_REASON"),
		record.Get("REC_SUB_HOUSING"),
		record.Get("REC_MED_ASSIST"),
		record.Get("REC_FOOD_STAMPS"),
		record.Get("REC_SUB_CC"),
		newUUID(),
		datafileID,
		int32(record.LineNumber),
	})
}

// SSP M5 Serializer - Adult-level data for closed SSP cases

func serializeSspM5(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		record.Get("RecordType"),
		record.Get("RPT_MONTH_YEAR"),
		record.Get("CASE_NUMBER"),
		record.Get("FAMILY_AFFILIATION"),
		record.Get("DATE_OF_BIRTH"),
		record.Get("SSN"),
		record.Get("RACE_HISPANIC"),
		record.Get("RACE_AMER_INDIAN"),
		record.Get("RACE_ASIAN"),
		record.Get("RACE_BLACK"),
		record.Get("RACE_HAWAIIAN"),
		record.Get("RACE_WHITE"),
		record.Get("SEX"),
		record.Get("REC_OASDI_INSURANCE"),
		record.Get("REC_FEDERAL_DISABILITY"),
		record.Get("REC_AID_TOTALLY_DISABLED"),
		record.Get("REC_AID_AGED_BLIND"),
		record.Get("REC_SSI"),
		record.Get("MARITAL_STATUS"),
		record.Get("RELATIONSHIP_HOH"),
		record.Get("PARENT_MINOR_CHILD"),
		record.Get("NEEDS_OF_PREGNANT_WOMAN"),
		record.Get("EDUCATION_LEVEL"),
		record.Get("CITIZENSHIP_STATUS"),
		record.Get("EMPLOYMENT_STATUS"),
		record.Get("AMOUNT_EARNED_INCOME"),
		record.Get("AMOUNT_UNEARNED_INCOME"),
		newUUID(),
		datafileID,
		int32(record.LineNumber),
	})
}

// SSP M6 Serializer - Aggregate data

func serializeSspM6(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		record.Get("RecordType"),
		record.Get("CALENDAR_QUARTER"),
		record.Get("RPT_MONTH_YEAR"),
		record.Get("SSPMOE_FAMILIES"),
		record.Get("NUM_2_PARENTS"),
		record.Get("NUM_1_PARENTS"),
		record.Get("NUM_NO_PARENTS"),
		record.Get("NUM_RECIPIENTS"),
		record.Get("ADULT_RECIPIENTS"),
		record.Get("CHILD_RECIPIENTS"),
		record.Get("NONCUSTODIALS"),
		record.Get("AMT_ASSISTANCE"),
		record.Get("CLOSED_CASES"),
		newUUID(),
		datafileID,
		int32(record.LineNumber),
	})
}

// SSP M7 Serializer - Stratum data

func serializeSspM7(record *parser.ParsedRecord, datafileID int32) [][]any {
	return singleRow([]any{
		record.Get("RecordType"),
		record.Get("CALENDAR_QUARTER"),
		record.Get("RPT_MONTH_YEAR"),
		record.Get("TDRS_SECTION_IND"),
		record.Get("STRATUM"),
		record.Get("FAMILIES_MONTH"),
		newUUID(),
		datafileID,
		int32(record.LineNumber),
	})
}
